File: src/main/java/com/hrms/hr_payroll_management_system/service/LeaveService.java
package com.hrms.hr_payroll_management_system.service;

import com.hrms.hr_payroll_management_system.common.pagination.PageResponse;
import com.hrms.hr_payroll_management_system.dto.request.leave.CreateLeaveRequest;
import com.hrms.hr_payroll_management_system.dto.request.leave.ReviewLeaveRequest;
import com.hrms.hr_payroll_management_system.dto.response.leave.LeaveRequestResponse;

public interface LeaveService {

    LeaveRequestResponse requestLeave(
            Long employeeId,
            CreateLeaveRequest request
    );

    PageResponse<LeaveRequestResponse> getByEmployeeId(
            Long employeeId,
            int page,
            int size
    );

    LeaveRequestResponse approve(
            Long id,
            ReviewLeaveRequest request
    );

    LeaveRequestResponse reject(
            Long id,
            ReviewLeaveRequest request
    );

    LeaveRequestResponse cancel(Long id);

    int accrueYearlyBalances(int year);
}

File: frontend/src/services/leave.service.js
import { leaveApi } from '../api/leave.api';

export const leaveService = {
  async listByEmployee(employeeId, page = 0, size = 10) {
    const res = await leaveApi.getByEmployeeId(employeeId, page, size);
    return res.data; // PageResponse<LeaveRequestResponse>
  },
  async create(employeeId, payload) {
    const res = await leaveApi.create(employeeId, payload);
    return res.data;
  },
  async approve(id, comment) {
    const res = await leaveApi.approve(id, { comment });
    return res.data;
  },
  async reject(id, comment) {
    const res = await leaveApi.reject(id, { comment });
    return res.data;
  },
  async cancel(id) {
    const res = await leaveApi.cancel(id);
    return res.data;
  },
};